jetty-security/src/main/java/org/eclipse/jetty/security/RoleInfo.java

```diff
@@ -140,26 +140,28 @@ public void combine(RoleInfo other)
     {
         if (other._forbidden)
             setForbidden(true);
-        else if (!other._checked) // TODO is this the right way around???
-            setChecked(true);
-        else if (other._isAnyRole)
-            setAnyRole(true);
-        else if (other._isAnyAuth)
-            setAnyAuth(true);
-        else if (!_isAnyRole)
+        else if (other._checked)
         {
-            for (String r : other._roles)
-            {
-                _roles.add(r);
-            }
+            setChecked(true);
+            if (other._isAnyRole)
+                setAnyRole(true);
+            else if (other._isAnyAuth)
+                setAnyAuth(true);
+            else if (!_isAnyRole)
+                _roles.addAll(other._roles);
```

Comment:
Shouldn't `_roles.addAll(other._roles)` be done in the `other._isAnyRole == true` case as well?

In ConstraintSecurityHandler.configureRoleInfo() it is done like that.
```java
if (mapping.getConstraint().isAnyRole())
{
    // * means matches any defined role
    for (String role : _roles)
    {
        ri.addRole(role);
    }
    ri.setAnyRole(true);
}
```